src/pool-cl/anti-sniping/CLAntiSniping.sol

```diff
@@ -0,0 +1,249 @@
+            hookDelta = BalanceDeltaLibrary.ZERO_DELTA;
+        }
+        return (this.afterRemoveLiquidity.selector, hookDelta);
+    }
+
+    /// @notice Handles logic before adding liquidity, enforcing position creation constraints.
+    /// @dev Records position creation block and ensures the position doesn't already exist or exceed the same block limit.
+    function beforeAddLiquidity(
+        address sender,
+        PoolKey calldata key,
+        ICLPoolManager.ModifyLiquidityParams calldata params,
+        bytes calldata
+    ) external override returns (bytes4) {
+        PoolId poolId = key.toId();
+        collectLastBlockInfo(poolId);
+        bytes32 positionKey = CLPosition.calculatePositionKey(sender, params.tickLower, params.tickUpper, params.salt);
+        LiquidityParams storage liqParams = positionKeyToLiquidityParams[positionKey];
+        liqParams.sender = sender;
+        liqParams.tickLower = params.tickLower;
+        liqParams.tickUpper = params.tickUpper;
+        liqParams.salt = params.salt;
+        positionKeyToLiquidityParams[positionKey] = liqParams;
+        if (positionCreationBlock[poolId][positionKey] != 0) revert PositionAlreadyExists();
+        if (positionsCreatedInLastBlock[poolId].length >= sameBlockPositionsLimit) revert TooManyPositionsInSameBlock();
+        positionCreationBlock[poolId][positionKey] = block.number;
+        positionsCreatedInLastBlock[poolId].push(positionKey);
+        return (this.beforeAddLiquidity.selector);
+    }
+
+    /// @notice Handles logic before removing liquidity, enforcing position lock duration and full withdrawal.
+    /// @dev Checks that the position lock duration has passed and disallows partial withdrawals.
+    function beforeRemoveLiquidity(
+        address sender,
+        PoolKey calldata key,
+        ICLPoolManager.ModifyLiquidityParams calldata params,
+        bytes calldata
+    ) external override returns (bytes4) {
+        PoolId poolId = key.toId();
+        collectLastBlockInfo(poolId);
+        bytes32 positionKey = CLPosition.calculatePositionKey(sender, params.tickLower, params.tickUpper, params.salt);
+        if (block.number - positionCreationBlock[poolId][positionKey] < positionLockDuration) revert PositionLocked();
+        CLPosition.Info memory info = poolManager.getPosition(poolId, sender, params.tickLower, params.tickUpper, params.salt);
+        if (int128(info.liquidity) + params.liquidityDelta != 0) revert PositionPartiallyWithdrawn();
+        return (this.beforeRemoveLiquidity.selector);
+    }
+
+    /// @notice Handles logic before a swap occurs.
+    /// @dev Collects fee information for positions created in the last processed block.
+    function beforeSwap(address, PoolKey calldata key, ICLPoolManager.SwapParams calldata, bytes calldata)
+        external
+        override
+        returns (bytes4, BeforeSwapDelta, uint24)
+    {
+        PoolId poolId = key.toId();
+        collectLastBlockInfo(poolId);
```

Comment:
Can u add some test cases so that we can compare the gas snapshot ? I am a bit curious that how many gas overhead will this function bring

Reply:
yes, will add testcase later